utilities/blob_db/blob_db_impl.cc

```diff
@@ -0,0 +1,2112 @@
+#include "rocksdb/convenience.h"
+#include "rocksdb/env.h"
+#include "rocksdb/iterator.h"
+#include "rocksdb/utilities/stackable_db.h"
+#include "rocksdb/utilities/transaction.h"
+#include "table/block.h"
+#include "table/block_based_table_builder.h"
+#include "table/block_builder.h"
+#include "table/meta_blocks.h"
+#include "util/crc32c.h"
+#include "util/file_reader_writer.h"
+#include "util/instrumented_mutex.h"
+#include "util/timer_queue.h"
+#include "utilities/transactions/optimistic_transaction_db_impl.h"
+#include "utilities/transactions/optimistic_transaction_impl.h"
+
+namespace {
+
+void extendTTL(rocksdb::blobstorage::ttlrange_t* ttl_range, uint32_t ttl) {
+  ttl_range->first = std::min(ttl_range->first, ttl);
+  ttl_range->second = std::max(ttl_range->second, ttl);
+}
+
+void extendTimestamps(rocksdb::blobstorage::tsrange_t* ts_range, uint64_t ts) {
+  ts_range->first = std::min(ts_range->first, ts);
+  ts_range->second = std::max(ts_range->second, ts);
+}
+
+void extendSN(rocksdb::blobstorage::snrange_t* sn_range,
+              rocksdb::SequenceNumber sn) {
+  sn_range->first = std::min(sn_range->first, sn);
+  sn_range->second = std::max(sn_range->second, sn);
+}
+}  // end namespace
+
+namespace rocksdb {
+
+namespace blobstorage {
+Random blob_rgen(static_cast<uint32_t>(time(nullptr)));
+
+void BlobDBFlushBeginListener::OnFlushBegin(DB* db, const FlushJobInfo& info) {
+  if (impl_) impl_->OnFlushBeginHandler(db, info);
+}
+
+WalFilter::WalProcessingOption BlobReconcileWalFilter::LogRecordFound(
+    unsigned long long log_number, const std::string& log_file_name,
+    const WriteBatch& batch, WriteBatch* new_batch, bool* batch_changed) {
+  return WalFilter::WalProcessingOption::kContinueProcessing;
+}
+
+bool blobf_compare_ttl::operator()(const std::shared_ptr<BlobFile>& lhs,
+                                   const std::shared_ptr<BlobFile>& rhs) const {
+  if (lhs->ttl_range_.first < rhs->ttl_range_.first) return true;
+
+  if (lhs->ttl_range_.first > rhs->ttl_range_.first) return false;
+
+  return lhs->BlobFileNumber() > rhs->BlobFileNumber();
+}
+
+void EvictAllVersionsFilter::Callback(int level, const Slice& key,
+                                      ValueType value_type,
+                                      const Slice& existing_value,
+                                      const SequenceNumber& sn,
+                                      bool is_new) const {
+  if (!is_new && value_type == CompactionFilter::kValue) {
+    BlockHandle handle;
+    uint64_t file_number;
+    bool succ =
+        BlobDBImpl::parseLSMValue(existing_value, &file_number, &handle);
+
+    if (succ) {
+      if (impl_->debug_level_ >= 3)
+        Log(InfoLogLevel::INFO_LEVEL, impl_->db_options_.info_log,
+            "CALLBACK COMPACTED OUT KEY: %s SN: %d "
+            "NEW: %d FN: %d OFFSET: %d SIZE: %d",
+            key.ToString().c_str(), sn, is_new, file_number, handle.offset(),
+            handle.size());
```

Comment:
size() will probably return size_t in which case would require a variable size specification which is defined in port files